stdlib/domain.ml

```diff
@@ -1,11 +1,90 @@
-type t = int
-external spawn : (unit -> unit) -> t = "caml_domain_spawn"
+module Raw = struct
+  (* Low-level primitives provided by the runtime *)
+  type t = private int
+  external critical_adjust : int -> unit
+    = "caml_ml_domain_critical_section"
+  external interrupt : t -> unit
+    = "caml_ml_domain_interrupt"
+  external wait : unit -> unit = "caml_ml_domain_yield"
+  external spawn : (unit -> unit) -> t
+    = "caml_domain_spawn"
+  external self : unit -> t
+    = "caml_ml_domain_id"
+end
+
+module Sync = struct
+  let critical_section f =
+    Raw.critical_adjust (+1);
+    match f () with
+    | x -> Raw.critical_adjust (-1); x
+    | exception ex -> Raw.critical_adjust (-1); raise ex
```

Comment:
Hmm. I'm relying on the idea that effects that act like exceptions are generally followed by actual exceptions (or will be, once FiberDeath is implemented). It's probably a bad idea to perform effects inside critical sections anyway, though.